### ISSUE TITLE
log entries are off by one

### ISSUE THREAD

**Issue by cyptus** [Zefau/ioBroker.nuki-extended#73]
I am using version `2.2.6` of the adapter with an nuki soft bridge adapter installation.
The entries in `nuki-extended.0.smartlocks.<lock>.logs` are always missing the last entry.

```typescript
on('nuki-extended.0.smartlocks.<lock>.logs', () => {
  const log = JSON.parse(getState('nuki-extended.0.smartlocks.<lock>.logs').val);
  console.log(log[0]);
});
```
is only logging the action before the last one.
I am able to validate this by manually checking the log entries from the object after any action.

How can i get the most recent action (with triggering user) triggered on the lock?

**Comment by Zefau** [comment]
I can't confirm this with my installation.

The log entries are retrieved from the Web API, which means that it matches the log you can find on https://web.nuki.io/. Could you please check if the most recent entries are available or missing there as well?

Furthermore, the up-to-dateness of the log depends on the refresh time you have set in the adapter settings for the Nuki Web API. What refresh frequency have you set?

**Comment by cyptus** [comment]
Thanks for your input @Zefau 
Refresh frequency has been to 0, but the log object has been refreshed / changed on actions to the nuki (always 1 entry off).
I changed the frequency to 5 seconds, now its working fine.

Where does the log changes come from while the setting is beeing set to 0?

**Comment by cyptus** [comment]
could the issue be https://github.com/Zefau/ioBroker.nuki-extended/blob/f655657a5f8689fb684a8f7c6eb2c7463cdf879a/nuki-extended.js#L386 ?
on callback by the bridge for any action, the adapter is refreshing the webApi as well.
But if there is any latency on the `bridge -> nuki web log`, the log may not have the most recent update by the bridge yet.
This would explain the off-by-one error.

**Comment by Zefau** [comment]
Yes, I guess that's it

**Comment by Zefau** [comment]
I quickly added a timeout for this case. Could you install the current Github version (no version number change for now) and verify the fix?

**Comment by cyptus** [comment]
kind of working - the nuki log is really weird.
I tried verifying the delay by stopping time from action to logs available in https://web.nuki.io/#/pages/activity-log (pressing "clear filter" continuously, as this fetchs the logs again).
The logs are sometimes delayed by 2-3 minutes - not sure if this has something todo with nukis android bridge.

is it currently possible to trigger `nukiWebApi.getSmartlockLogs` by script?

**Comment by Zefau** [comment]
So it is fixed with regards to the ioBroker adapter, but not fixed with regards to Nuki Bridge vs. Web API ?
I assigned `foreign issue` to this. Let's see what answer you get in the Nuki Developer forum.

**Comment by cyptus** [comment]
Yes, it cannot be fixed within this adapter as the nuki web logs are delayed themself.
The software bridge is not able to provide logs recording to the bridge api documentation, not sure about the details the hardware bridge could provide.

Hopefully nuki will provide a solution from their side, otherwise this adapter should mention the log delay within the documentation and maybe provide some kind of trigger object to fetch logs from web api again.
This way scripts would be able to handle the unknown delay to get details about the last action from a bridge-callback (e.g. the user which triggered the action).

Also nuki notification callbacks could be a solution for this, but they are still in beta.
I will keep this up2date if there is any response by nuki.

**Comment by mcm1957** [comment]
no more infomation since a long time. Lets close.
Please open a new issue if anything should be fixed / improved.
